Name: Andres Figueroa

Email: andresfigueroa@brandeis.edu

Description: The goal of this project is to build a model that predicts whether a pokemon is Legendary or not based on
its stats and other features.

        * Machine Learning Type: Binary Classification (True or False, Legendary or Not)

        * What will determine success is accuracy and precision ( we don't want to falsely classify weak pokemon as 
          legendary)


# Loading and Exploring Data

In [1]:
import pandas as pd

df = pd.read_csv('pokemon.csv')

print(df.columns)
print(df.head())

Index(['abilities', 'against_bug', 'against_dark', 'against_dragon',
       'against_electric', 'against_fairy', 'against_fight', 'against_fire',
       'against_flying', 'against_ghost', 'against_grass', 'against_ground',
       'against_ice', 'against_normal', 'against_poison', 'against_psychic',
       'against_rock', 'against_steel', 'against_water', 'attack',
       'base_egg_steps', 'base_happiness', 'base_total', 'capture_rate',
       'classfication', 'defense', 'experience_growth', 'height_m', 'hp',
       'japanese_name', 'name', 'percentage_male', 'pokedex_number',
       'sp_attack', 'sp_defense', 'speed', 'type1', 'type2', 'weight_kg',
       'generation', 'is_legendary'],
      dtype='object')
                     abilities  against_bug  against_dark  against_dragon  \
0  ['Overgrow', 'Chlorophyll']          1.0           1.0             1.0   
1  ['Overgrow', 'Chlorophyll']          1.0           1.0             1.0   
2  ['Overgrow', 'Chlorophyll']          1.0         

Our label will be 'is_legendary'

# Cleaning and Preparing Data

In [7]:
to_drop = ['abilities', 'name', 'percentage_male', 'pokedex_number', 'generation', 'type1', 'type2', 'japanese_name', 'base_egg_steps', 'classfication'] # Some features that I think are irrelevant to the problem
df = df.drop(columns = to_drop, errors = 'ignore')

to_keep = []
for col in df.columns:
    if 'against' not in col: # I want to get rid of all features with 'against' because their weakness tell us nothing about whether they are legendary or not
        to_keep.append(col)

df = df[to_keep]
#print(df.dtypes)
df['capture_rate'] = pd.to_numeric(df['capture_rate'], errors='coerce')  # safely convert to float or int#print(df.dtypes)
df = df.dropna()

df.columns

Index(['attack', 'base_happiness', 'base_total', 'capture_rate', 'defense',
       'experience_growth', 'height_m', 'hp', 'sp_attack', 'sp_defense',
       'speed', 'weight_kg', 'is_legendary'],
      dtype='object')

Explanation:
The variable 'to_drop' has a bunch of features that are irrelevant to the ML problem (whether a pokemon is Legendary or not), so they're going to be dropped.

I dropped the all columns (features) with 'against' because their weakness tell us nothing about whether they are legendary or not.

I decided to drop the feature 'base_egg_steps' because legendary pokemon don't hatch from eggs and this will essentially cause a data leakage, giving the model the answers rather than learning anything.

'capture_rate' was of the object type, specifically a string. So, I converted it to a numerical value

# Defining Label and Features

In [3]:
y = df['is_legendary']
X = df.drop(columns = 'is_legendary')

# Splitting the Data, Train and Test

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1234, stratify = y)

I chose to stratify to make sure that the class distribution stays the same in both the training and test datasets. Legendary pokemon are rare, so I wanna make sure that each dataset actually has legendary pokemon in it.

# Training a Decision Tree

In [5]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report

dt = DecisionTreeClassifier(max_depth=4, random_state=42)
dt.fit(X_train, y_train)

y_pred = dt.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.9871794871794872
              precision    recall  f1-score   support

           0       1.00      0.99      0.99       142
           1       0.88      1.00      0.93        14

    accuracy                           0.99       156
   macro avg       0.94      0.99      0.96       156
weighted avg       0.99      0.99      0.99       156



The values above are performance metrics ("accuracy", "precision", "recall", "f1-score"), numbers that help us measure how well a model's predictions were. 

Accuracy tells us how many of the total predictions were correct out of all the predictions (.9871).

Precision tells us how many the the predicted legendary pokemon were actually legendary (0.88).

Recall tells us how many legendary pokemon were found out of all the actual legendary pokemon (1.00)

F1-Score gives us the balance between precision and recall

Support is the number of legendary pokemon in the set.

The model catches all legendary pokemon with a 100% recall. The model is very accurate (0.9871), but there are some false positives (precision at 0.88).